src/test/kotlin/org/rust/ide/console/RsConsoleFoldingTestBase.kt

```diff
@@ -0,0 +1,112 @@
+import com.intellij.openapi.util.Disposer
+import org.rust.RsTestBase
+import org.rust.fileTreeFromText
+
+open class RsConsoleFoldingTestBase : RsTestBase() {
+    private var myConsole: ConsoleViewImpl? = null
+
+    public override fun setUp() {
+        super.setUp()
+
+        val console = createConsole()
+        console.setUpdateFoldingsEnabled(true)
+        myConsole = console
+
+        val point: ExtensionPoint<ConsoleFolding> = ConsoleFolding.EP_NAME.point
+        point.registerExtension(RsConsoleFolder(), testRootDisposable)
+    }
+
+    public override fun tearDown() {
+        try {
+            Disposer.dispose(myConsole!!)
+        } catch (e: Throwable) {
+            addSuppressedException(e)
+        } finally {
+            super.tearDown()
+        }
+    }
+
+    protected fun doFoldingTest(tree: String, text: String) {
+        fileTreeFromText(tree).create()
+
+        val (input, expectedFolds) = createConsoleFoldInput(text)
+
+        val console = myConsole!!
+        console.print(input, ConsoleViewContentType.NORMAL_OUTPUT)
+        console.flushDeferredText()
+
+        val editor = console.editor
+        val regions = editor.foldingModel.allFoldRegions
+
+        val expectedText = renderTextWithFolds(input, expectedFolds)
+        val actualText = renderTextWithFolds(input,
+            regions.map { Fold(it.placeholderText, it.startOffset, it.endOffset) })
+        assertEquals(expectedText, actualText)
+    }
+
+    private fun createConsole(): ConsoleViewImpl {
+        val console = RsConsoleView(project)
+        console.component // initialize component
+        val processHandler: ProcessHandler = NopProcessHandler()
+        processHandler.startNotify()
+        console.attachToProcess(processHandler)
+        return console
+    }
+
+    companion object {
+        private fun renderTextWithFolds(text: String, folds: List<Fold>): String = buildString {
+            var offset = 0
+            folds.forEach { fold ->
+                append(text.substring(offset until fold.startOffset))
+                append(fold.text)
+                offset = fold.endOffset
+            }
+            append(text.substring(offset))
+        }
+
+        private fun createConsoleFoldInput(text: String): Pair<String, List<Fold>> {
+            val expectedFolds = mutableListOf<Fold>()
+            val input = buildString {
+                var offset = 0
+                var currentText: String? = null
+                var currentStartOffset: Int? = null
+                for (line in text.trimIndent().lineSequence()) {
+                    when {
+                        line.startsWith("//foldstart") -> {
+                            currentText = line.removePrefix("//foldstart ").trim()
+                            currentStartOffset = offset
+                        }
+                        line.startsWith("//foldend") -> {
+                            assertNotNull(currentText)
```

Comment:
Do you really need this check? You use `!!` on next line anyway. Note, you don't do it for `currentStartOffset` variable. If you still want to do it, let's better use `check(currentText  != null)`. It uses kotlin contracts under the hood and it allows compiler to make smart cast to not null value after the check

Reply:
It's just a sanity check if the test input is correct. I used `check`, as you hinted.